### PR TITLE
Fixes crash due to accessing selection after graph_scene is deleted #218

### PR DESCRIPTION
**Pull request by valleyofblackpanther**
### Summary 
The `MainWindow `class is where I integrated the cleanup routine for the `GraphScene`. `MainWindow` has already a `closeEvent` method. This is where we call the shutdown method of each `GraphScene` instance before the application closes. However, the `GraphScene` instances are not directly managed by `MainWindow` but rather by the panels like `GraphEditPanel`, `ProofPanel`, and `RulePanel` which are derived from `BasePanel`. Each of these panels contains an instance of `GraphScene`.

Here's the cleanup implementation:
1. Add a cleanup method in the `BasePanel` class.
2. Call this cleanup method for each panel in the MainWindow's closeEvent.

Firslty, I added a `shutdown` method to the `GraphScene` class. Then, added a corresponding method in the `BasePanel` class that calls `GraphScene.shutdown()`
Next, I ensured each panel subclassing `BasePanel` class provides its own shutdown method. 
Finally, I modified the `closeEvent` of the `MainWindow` class to shutdown on each panel before closing. 

This method ensures that each `GraphScene` is properly cleaned up before the main window closes.

### CODE REVIEW

File: zxlive/edit_panel.py [hunk 1]
```diff
@@ -106,3 +106,7 @@ def _input_circuit(self) -> None:
             cmd = UpdateGraph(self.graph_view, new_g)
             self.undo_stack.push(cmd)
             self.graph_scene.select_vertices(new_verts)
+
+    def shutdown(self):
```
Comment (RazinShaikh) [inline zxlive/edit_panel.py:110]: Why do you need this method here in `edit_panel` if it is already implemented in `base_panel`?

File: zxlive/proof_panel.py [hunk 2]
```diff
@@ -421,6 +421,10 @@ def _refresh_rewrites_model(self) -> None:
         # TODO: Right now this calls for every single vertex selected, even if we select many at the same time
         self.graph_scene.selectionChanged.connect(model.update_on_selection)
 
+    def shutdown(self):
```
Comment (RazinShaikh) [inline zxlive/proof_panel.py:424]: Why do you need this method here in `proof_panel` if it is already implemented in `base_panel`?

File: zxlive/rule_panel.py [hunk 3]
```diff
@@ -80,3 +80,7 @@ def get_rule(self) -> CustomRule:
                           self.graph_scene_right.g,
                           self.name_field.text(),
                           self.description_field.text())
+    
+    def shutdown(self):
```
Comment (RazinShaikh) [inline zxlive/rule_panel.py:84]: Why do you need this method here in `rule_panel` if it is already implemented in `base_panel`?

File: zxlive/graphscene.py [hunk 4]
```diff
@@ -195,6 +195,20 @@ def select_all(self) -> None:
...
+    def shutdown(self):
+
+        """
+        Perform necessary cleanup to avoid RunTimeError when the application closes.
+        Call this method before the application closes. 
+        """        
+
+        #Clear all items from the scene to ensure no dangling references remain
+        self.clear()
+
```
Comment (RazinShaikh) [inline zxlive/graphscene.py:207]: Do you mind cleaning up some extra empty lines in this function?

File: zxlive/dialogs.py [hunk 5]
```diff
@@ -76,9 +76,9 @@ class ImportRuleOutput:
     file_path: str
     r: CustomRule
 
-def show_error_msg(title: str, description: Optional[str] = None) -> None:
+def show_error_msg(title: str, description: Optional[str] = None, parent: Optional[QWidget] = None) -> None:
```
Comment (RazinShaikh) [inline zxlive/dialogs.py:79]: The changes in this file regarding the parent are related to a separate issue. Do you mind splitting this into a separate PR?

**Review by RazinShaikh — CHANGES_REQUESTED** [review]
Hi Kaushik, thank you for the PR. Could you please make the requested changes before we can merge this?

File: zxlive/graphscene.py (hunk 4 above)
Comment (valleyofblackpanther) [reply zxlive/graphscene.py]: Hi, @RazinShaikh I have made the suggested changes, here is the [commit](https://github.com/Quantomatic/zxlive/pull/230/commits/cb113bde6d0ba31fad648426a89c990a0b4a8d2f)

**Comment by valleyofblackpanther** [comment]
Hi, @RazinShaikh I have made the suggested changes, here is the [commit](https://github.com/Quantomatic/zxlive/pull/230/commits/cb113bde6d0ba31fad648426a89c990a0b4a8d2f).

**Comment by dlyongemallo** [comment]
I wonder if it wouldn't be simpler and more future-proof to simply clear the node selection when the user asks to close the app?

**Comment by RazinShaikh** [comment]
@dlyongemallo that actually makes more sense

File: zxlive/mainwindow.py [hunk 6]
```diff
@@ -223,6 +223,12 @@ def active_panel(self) -> Optional[BasePanel]:
 
 
     def closeEvent(self, e: QCloseEvent) -> None:
+        #Call shutdown on all tabs before closing
```
Comment (dlyongemallo) [inline zxlive/mainwindow.py:226]: I think you may be overthinking this. I think all that is needed to fix the issue is to add a single line in `close_tab` (just above line 329 `self.tab_widget.removeTab(i)`), like:
`widget.graph_scene.clearSelection()`

Please try to see if that works.

**Comment by dlyongemallo** [comment]
> Hi Kaushik, thank you for the PR. Could you please make the requested changes before we can merge this?

Thanks for the PR, Kaushik. Can you make the suggested change and add a test to confirm it works? I think that should suffice.
